esphome/components/lc709203f/lc709203f.cpp

```diff
@@ -0,0 +1,302 @@
+static const uint8_t LC709203F_CELL_VOLTAGE = 0x09;
+static const uint8_t LC709203F_CURRENT_DIRECTION = 0x0A;
+static const uint8_t LC709203F_APA = 0x0B;
+static const uint8_t LC709203F_APT = 0x0C;
+static const uint8_t LC709203F_RSOC = 0x0D;
+static const uint8_t LC709203F_ITE = 0x0F;
+static const uint8_t LC709203F_IC_VERSION = 0x11;
+static const uint8_t LC709203F_CHANGE_OF_THE_PARAMETER = 0x12;
+static const uint8_t LC709203F_ALARM_LOW_RSOC = 0x13;
+static const uint8_t LC709203F_ALARM_LOW_CELL_VOLTAGE = 0x14;
+static const uint8_t LC709203F_IC_POWER_MODE = 0x15;
+static const uint8_t LC709203F_STATUS_BIT = 0x16;
+static const uint8_t LC709203F_NUMBER_OF_THE_PARAMETER = 0x1A;
+
+static const uint8_t LC709203F_POWER_MODE_ON = 0x0001;
+static const uint8_t LC709203F_POWER_MODE_SLEEP = 0x0002;
+
+static const uint8_t LC709203F_STATE_INIT = 0x01;
+static const uint8_t LC709203F_STATE_RSOC = 0x02;
+static const uint8_t LC709203F_STATE_TEMP_SETUP = 0x03;
+static const uint8_t LC709203F_STATE_NORMAL = 0x00;
+
+// The number of times to retry an I2C transaction before giving up. In my experience,
+//  10 is a good number here that will take care of most bus issues that require retry.
+static const uint8_t LC709203F_I2C_RETRY_COUNT = 10;
+
+void Lc709203f::setup() {
+  // Note: The setup implements a small state machine. This is because we want to have
+  //  delays before and after sending the RSOC command. The full init process should be:
+  //       INIT->RSOC->TEMP_SETUP->NORMAL
+  //  The setup() function will only perform the first part of the initialization process.
+  //  Assuming no errors, the whole process should occur during the setup() function and
+  //  the first two calls to update(). After that, the part should remain in normal mode
+  //  until a device reset.
+  //
+  //  This device can be picky about I2C communication and can error out occasionally. The
+  //  get/set register functions impelment retry logic to retry the I2C transactions. The
+  //  initialization code checks the return code from those functions. If they don't return
+  //  NO_ERROR (0x00), that part of the initialization aborts and will be retried on the next
+  //  call to update().
+  ESP_LOGCONFIG(TAG, "Setting up LC709203F...");
+
+  this->state_ = LC709203F_STATE_INIT;
+  // Set power mode to on. Note that, unlike some other similar devices, in sleep mode the IC
+  //  does not record power usage. If there is significant power consumption during sleep mode,
+  //  the pack RSOC will likely no longer be correct. Because of that, I do not implement
+  //  sleep mode on this device.
+  if (this->set_register_(LC709203F_IC_POWER_MODE, LC709203F_POWER_MODE_ON) == i2c::NO_ERROR) {
+    if (this->set_register_(LC709203F_APA, this->apa_) == i2c::NO_ERROR) {
+      if (this->set_register_(LC709203F_CHANGE_OF_THE_PARAMETER, this->pack_voltage_) == i2c::NO_ERROR) {
+        this->state_ = LC709203F_STATE_RSOC;
+      }
+    }
+  }
```

Comment:
Nothing specifically wrong here, but...how many levels deep are you going to stack these? 😅
```suggestion
  if (this->set_register_(LC709203F_IC_POWER_MODE, LC709203F_POWER_MODE_ON) != i2c::NO_ERROR) {
    return;
  }
  if (this->set_register_(LC709203F_APA, this->apa_) != i2c::NO_ERROR) {
    return;
  }
  if (this->set_register_(LC709203F_CHANGE_OF_THE_PARAMETER, this->pack_voltage_) != i2c::NO_ERROR) {
    return;
  }
  this->state_ = LC709203F_STATE_RSOC;
```